### PR TITLE
File header code fix should only replace the copyright element

### LINKED ISSUE #1434: File header code fix should only replace the copyright element
An XML file header may contain additional elements, including the following:

1. Horizontal bars (the series of `-----------------` characters)
2. A `<summary>` element
3. Other XML elements
4. Other text outside of XML elements

## Expected behavior

When the file header code fix corrects the `<copyright>` element, each of these other elements should be preserved.

## Actual behavior

When the file header code fix corrects the `<copyright>` element, the entire header is replaced (each of the items is lost).

### PR DESCRIPTION
**Pull request by peteri**
Fixes for SA1634 & SA1636 to deal with other copyright header elements. If the header is already valid then this only replaces the <copyright> tag contents. Otherwise it does the previous behaviour of replacing the whole header.

Also removed trailing spaces on blank lines in copyright text.

Fixes #1434

### CODE REVIEW

**Comment by sharwell** [comment]
I'm looking forward to this one!

**Comment by peteri** [comment]
Don't think they duplicate (although I did get merge grief)
Currently the fixer doesn't deal with #1432 but now the other code is in place I'm happy to have a look.

**Comment by peteri** [comment]
Realised I had some unreachable code, so removed it and cleaned up the logic, also removed an inheritdoc xml comment tag that had appeared somehow.

**Comment by vweijsters** [comment]
Valid XML file headers are not handled (See #1432)

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs [hunk 1]
```diff
@@ -106,8 +140,32 @@ private static SyntaxNode ReplaceHeader(Document document, SyntaxNode root, Styl
...
+            var newHeaderTrivia = CreateNewHeader(document.Name, settings, newLineText);
+            if (!isMalformedHeader && copyrightTriviaIndex.HasValue)
+            {
+                // Replace copyright element in place.
+                return root.WithLeadingTrivia(trivia.ReplaceRange(trivia[copyrightTriviaIndex.Value], newHeaderTrivia));
+            }
+            else
+            {
+                // Add blank line if we don't already have comments at top of file.
+                if ((trivia.Count == 0) || (trivia[0].Kind() != SyntaxKind.SingleLineCommentTrivia))
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: `trivia[0].Kind() != SyntaxKind.SingleLineCommentTrivia` can be replaced by `!trivia[0].IsKind(SyntaxKind.SingleLineCommentTrivia)`

**Comment by vweijsters** [comment]
I'm missing test cases for a single line copyright block. For example:
```C#
// <copyright file=""Test0.cs"" company=""FooCorp""> Copyright (c) FooCorp. All rights reserved.</copyright>
// <author>John Doe</author>

namespace Test
{
}
```

**Comment by vweijsters** [comment]
:thought_balloon: The code is currently very difficult to analyze for correctness because there is no separation between cases where an XML header is expected or not, depending on the settings (StyleCopSettings.DocumentationRules.XmlHeader). My proposal would be to have two `ReplaceHeader` implementations, one for XML based headers and one for non-XML headers.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs [hunk 2]
```diff
@@ -56,48 +57,81 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                            var closingTag = triviaLine.ToFullString().Contains("</copyright>") ||
+                                (openingTag && triviaLine.ToFullString().Trim().EndsWith("/>"));
+                            if (openingTag)
+                            {
+                                inCopyright = !closingTag;
+                                copyrightTriviaIndex = i;
+                            }
+                            else
+                            {
+                                if (inCopyright)
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: :thought_balloon: Reformatting to `else if (inCopyright)` would improve readability

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs [hunk 3]
```diff
@@ -56,48 +57,81 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
             var settings = document.Project.AnalyzerOptions.GetStyleCopSettings();
 
             var fileHeader = FileHeaderHelpers.ParseFileHeader(root);
-            var newSyntaxRoot = fileHeader.IsMissing ? AddHeader(document, root, document.Name, settings) : ReplaceHeader(document, root, settings);
+            var xmlFileHeader = FileHeaderHelpers.ParseXmlFileHeader(root);
+            var newSyntaxRoot = fileHeader.IsMissing ? AddHeader(document, root, document.Name, settings) : ReplaceHeader(document, root, settings, xmlFileHeader.IsMalformed);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: The code above should select the correct parsing method based on the `XmlHeader` setting.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs [hunk 4]
```diff
@@ -106,8 +140,32 @@ private static SyntaxNode ReplaceHeader(Document document, SyntaxNode root, Styl
                 }
             }
 
+            // Remove copyright lines in reverse order
+            removalList.Reverse();
+            foreach (var triviaLine in removalList)
+            {
+                trivia = trivia.RemoveAt(triviaLine);
+            }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: This can be implemented more efficiently using an for loop with a decreasing indexer.

**Comment by vweijsters** [comment]
The `AddHeader` method retrieves the correct newline to use and the adds `SyntaxFactory.CarriageReturnLineFeed` instead of the retrieved newline setting.

(Pre-existing issue)

**Comment by vweijsters** [comment]
There is an issue if the current file header has indentation. The copyright part would end up without any indentation, while the indentation for the rest of the header is preserved.

The previously implementation would remove all indentation, which is not correct, but preferable to having a mixed mode.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs [hunk 5]
```diff
@@ -106,8 +140,32 @@ private static SyntaxNode ReplaceHeader(Document document, SyntaxNode root, Styl
...
+            {
+                // Add blank line if we don't already have comments at top of file.
+                if ((trivia.Count == 0) || (trivia[0].Kind() != SyntaxKind.SingleLineCommentTrivia))
+                {
+                    newHeaderTrivia = newHeaderTrivia.Add(SyntaxFactory.CarriageReturnLineFeed);
+                }
+
+                // Insert header at top of the file.
+                return root.WithLeadingTrivia(newHeaderTrivia.Add(SyntaxFactory.CarriageReturnLineFeed).AddRange(trivia));
+            }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: Line 163 and 167 should use the newLineText, instead of `SyntaxFactory.CarriageReturnLineFeed`

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs (hunk 3 above)
Comment (peteri) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: I'll review this once I've had a look at doing the multiline comments, but I think you're correct and two lots of code would be cleaner.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs (hunk 1 above)
Comment (peteri) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: Didn't spot that (will change)

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs (hunk 4 above)
Comment (peteri) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: Okay I'll change that.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs (hunk 5 above)
Comment (peteri) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: Re: The indentation @sharwell in #1356 goes with the view that spaces should be added in the settings to the copyright text, I tend to agree as it gives a decent level of control and is at least predictable (whereas a magic autoindenter can go wrong easily).

Re Lines 163 and 167 this needs to be SyntaxFactory.EndOfLine(newLineText);
Is there an issue already raised for the other instances in the code base?
Feels like something that really needs an helper method somewhere.

Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: I added a helper method for the end of line in #1511, but that is not merged yet.

Regarding the indentation I think there is an unclarity. I was talking about leading spaces before the start of the single line comment and that cannot be handled by additional spaces in the settings.

Comment (peteri) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/FileHeaderCodeFixProvider.cs]: Ah see what you mean now. You're talking about 

```C#
   // <author>Peter</author>
   // <copyright>FooBar</copyright>
   // <summary>Stuff<summary>
```

with the old code would have been turned into

```C#
// <copyright>FooBar</copyright>
```

but now looks like this

```C#
   // <author>Peter</author>
// <copyright>FooBar</copyright>
   // <summary>Stuff<summary>
```

**Comment by sharwell** [comment]
@peteri Please let me know if you get stuck on this. As I'm sure you're aware, this is a very important bug fix to get out to users. :+1: I checked with @vweijsters today and this is the current status as we currently see it:

> #1505 still has issues with multiline comment headers as described in #1432. I consider that blocking

**Comment by peteri** [comment]
I had an attempt yesterday, but made too many changes at once between running tests and made a complete hash of it.
I will have a think while I'm at work and get the implementation done this evening (I'm in the UK). I'll let you know how far I get.

**Comment by peteri** [comment]
Looking pretty good for finishing tomorrow.

**Comment by peteri** [comment]
Could probably do with some more edge case tests but should be there now.

**Comment by vweijsters** [comment]
:+1: Most of it looks good.  Multiline comment trivia are handled properly now. There is one remark left that I would prefer to be addressed.

**Comment by peteri** [comment]
Hmm the single line comment check is so that this code 

    // <summary>Foo</summary>

comes out like this after the fix is run (I've used a single line comment for the copyright here so it's fairly brief)

```
// <copyright>Single line</copyright>
// <summary>Foo</summary>
```

but
```
#define STUFF
```

Looks like this after the fix

```
// <copyright>Single line</copyright>

#define STUFF
```

So for the multi line comment case (which can't be valid xml otherwise we would have gone down the other path)

```
/************** 
*  DON'T MOVE *
**************/
```

Will end up looking like this option (A):

```
// <copyright>Single line</copyright>

/************** 
*  DON'T MOVE *
**************/
```

Rather than option (B) below
```
// <copyright>Single line</copyright>
/************** 
*  DON'T MOVE *
**************/
```

My personal preference is for option A (a happy accident of the code) but I'm really not that fussy.

I'm out after work this evening but I can make a change tomorrow (Thursday) if it's wanted.

**Comment by sharwell** [comment]
Thanks for all the hard work on this y'all! :smile:
